Revert "Redefine scheduler to set learning rate using recursive formula" #14010

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21463 Revert "Redefine scheduler to set learning rate using recursive formula" #14010**

This revert was done by hand, by replacing the recursive get_lr definitions with
the non-recursive ones and then deleting tests that failed.  I did NOT delete
the "trivial" legacy implementation tests; they might still be useful if we
attempt this patch again.

Fixes #20527

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D15747426](https://our.internmc.facebook.com/intern/diff/D15747426)

ezyang committed:
Revert "Redefine scheduler to set learning rate using recursive formula" #14010

This revert was done by hand, by replacing the recursive get_lr definitions with
the non-recursive ones and then deleting tests that failed.  I did NOT delete
the "trivial" legacy implementation tests; they might still be useful if we
attempt this patch again.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
cc @jasam-sheja @jafermarq

@soumith We should consider including this revert in a point release, if we decide to do it.

ezyang commented:
@pytorchbot retest this please

ezyang committed:
Update on "Revert "Redefine scheduler to set learning rate using recursive formula" #14010"

Revert "Redefine scheduler to set learning rate using recursive formula" #14010

This revert was done by hand, by replacing the recursive get_lr definitions with
the non-recursive ones and then deleting tests that failed.  I did NOT delete
the "trivial" legacy implementation tests; they might still be useful if we
attempt this patch again.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 21463 gh/ezyang/157/head

ezyang committed:
Update on "Revert "Redefine scheduler to set learning rate using recursive formula" #14010"

Revert "Redefine scheduler to set learning rate using recursive formula" #14010

This revert was done by hand, by replacing the recursive get_lr definitions with
the non-recursive ones and then deleting tests that failed.  I did NOT delete
the "trivial" legacy implementation tests; they might still be useful if we
attempt this patch again.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 21463 gh/ezyang/157/head

ezyang commented:
@fmassa  says this will break torchvision code if we revert it.

fmassa commented:
I can update torchvision to use my own lr_schedulers to bypass a potential revert, like what I did in [maskrcnn-benchmark in the past](https://github.com/facebookresearch/maskrcnn-benchmark/blob/c8eff2ca2d5c1c14393f1a1b21853895eaacfd51/maskrcnn_benchmark/solver/lr_scheduler.py#L7-L9), but it would be great to have a way of composing lr schedulers natively supported in PyTorch.

zou3519 reviewed (commented):
there are changes to third_party/fbgemm

ezyang committed:
Update on "Revert "Redefine scheduler to set learning rate using recursive formula" #14010"

Revert "Redefine scheduler to set learning rate using recursive formula" #14010

This revert was done by hand, by replacing the recursive get_lr definitions with
the non-recursive ones and then deleting tests that failed.  I did NOT delete
the "trivial" legacy implementation tests; they might still be useful if we
attempt this patch again.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 21463 gh/ezyang/157/head

zou3519 commented on test/test_optim.py:
```diff
@@ -744,141 +744,6 @@ def test_reduce_lr_on_plateau8(self):
                                       threshold=0.1, patience=5, cooldown=5)
         self._test_reduce_lr_on_plateau(scheduler, targets, metrics, epochs)
 
-    def test_compound_step_and_multistep_lr(self):
```
I'm surprised that these are the only tests that failed; #14010 added a lot of things into test_optim.py

ezyang replied:
Yeah, the other tests were just comparing against reference implementations (which still works; it's just a less interesting test.)
